### PR TITLE
[MRG] fix support for categorical variables

### LINKED ISSUE #97: Support for categorical parameters seems to be broken
```python
def bench1(x):
    return np.asscalar(np.asarray(x))

def bench2(x):
    return np.asscalar(np.asarray(x, dtype=np.int))

bench1([1])
1
bench2(["1"])
1

from skopt import forest_minimize
# Works
forest_minimize(bench1, ((1.0, 4.0),))
# Fails
forest_minimize(bench2, (("1", "2", "3", "4"),))

# Works
gp_minimize(bench1, ((1.0, 4.0),), maxiter=5)
# Fails
gp_minimize(bench2, (("1", "2", "3", "4"),))
```

### PR DESCRIPTION
**Pull request by glouppe**
This fixes #97, modulo the fact that the function to minimize actually returns a float.

The main issue was that `Xi = np.zeros((maxiter, space.n_dims))` was forcing entries to be cast to floats, no matter the type of the dimensions. Samples are now stored in a list instead, as was done in `gp_minimize`.

### CODE REVIEW

**Comment by betatim** [comment]
Merge commits, woah! Can I be pedantic: rebase please 🙏

**Comment by glouppe** [comment]
F*, i messed up the history now...

**Comment by glouppe** [comment]
> Merge commits, woah! Can I be pedantic: rebase please :pray:

Done

**Comment by glouppe** [comment]
Hmm some tests are failing, but this has nothing to do with what I changed... ?!

**Comment by betatim** [comment]
Ergh, for some of the surrogate models some of the convergence tests fail. Have you tried this branch before and after applying 57025cc (or does it not happen at all locally?)

Is the right conclusion from this that somewhere we are getting non reproducible random numbers? As in we changed some code and now get a different sequence?

File: skopt/tree_opt.py [hunk 1]
```diff
@@ -39,38 +33,39 @@ def _tree_minimize(func, dimensions, base_estimator, maxiter,
         raise ValueError("Total number of iterations set by maxiter has to"
                          " be larger or equal to n_start.")
 
-    Xi[:n_start] = space.rvs(n_samples=n_start, random_state=rng)
-    yi[:n_start] = [func(xi) for xi in Xi[:n_start]]
-    i = np.argmin(yi[:n_start])
-    best_y = yi[i]
-    best_x = Xi[i]
+    Xi = space.rvs(n_samples=n_start, random_state=rng)
+    yi = [func(x) for x in Xi]
```
Comment (MechCoder) [inline skopt/tree_opt.py]: nitpick: We can still preallocate `y`

**Comment by MechCoder** [comment]
> some tests are failing, but this has nothing to do with what I changed

It has "almost" nothing to do with what you changed :-D . You need to revert `next_x = X[np.argmin(values)]` to `next_x = X[np.argmax(values)]` as before. We would like to maximize the expected improvement, which minimizes the objective.

**Comment by MechCoder** [comment]
Thanks for addressing this, +1 pending comments (and hopefully Travis)

**Pull request merged**